Updates artist page design to match vision

sweir27 opened the pull request:
This PR:
- Removes the `shows`, `cv`, and `articles` tabs in favor of putting them on the overview page
- Removes the artworks filter from the overview tab and puts it on a separate `/works-for-sale` page
- Adds some redirect logic:
  - If the artist has no content (bio, insights, etc.), redirect to the `/works-for-sale` tab and show the zero state
  - Adds a new route: `/works-for-sale`
  - All existing sub routes (i.e. `/artist/andy-warhol/cv`) re-route to the root

There are a few things that aren't quite working:
- ~Filters 😭~
- ~An error which I suspect is also related to some incorrect filter setup: ~
<img width="1174" alt="image" src="https://user-images.githubusercontent.com/2081340/70960281-99504780-204c-11ea-895f-eba4954a8cf4.png">

Turns out the filters were broken due to a local issue with my relay cache, and that warning that I was seeing is expected (cc @mzikherman ).

I will probably set this up as a review app for easier UI testing, but other than those issues above I believe it's good to go!

### Lots of content
<img width="1680" alt="image" src="https://user-images.githubusercontent.com/2081340/70960362-d1f02100-204c-11ea-87c1-821eb0af5d9e.png">

<img width="1680" alt="image" src="https://user-images.githubusercontent.com/2081340/70960185-5c845080-204c-11ea-9a47-bc8a5c5bd179.png">

### Some content
<img width="1680" alt="image" src="https://user-images.githubusercontent.com/2081340/70960230-72921100-204c-11ea-9d18-5d486802e165.png">

<img width="1680" alt="image" src="https://user-images.githubusercontent.com/2081340/70960168-49718080-204c-11ea-9b1a-bbfbce738354.png">


### No content
<img width="1680" alt="image" src="https://user-images.githubusercontent.com/2081340/70960149-3bbbfb00-204c-11ea-850c-079b1e8df5e0.png">

damassi commented:
@sweir27 - I've noticed that error due to mismatches between case / form wen passing query params and / or route params (e.g., `path: '/artist/:id'`. The casing is often passed verbatim over to elastic search. 

This PR shows what I mean:
- [Here](https://github.com/artsy/reaction/pull/2820/files#diff-6d6dfa2da310c5900f2ced293fc897aeL32), where `artistID` -> `artist_id`. 
- Because [here](https://github.com/artsy/reaction/pull/2820/files#diff-1d31f38d87470c693f29382aad7cbf1aR104), our filter endpoint expects `artist_id` to be passed. 

This took me a sec to figure out when I was refactoring the filter. Down to help brainstorm.

damassi commented on src/Apps/Artist/routes.tsx:
```diff
@@ -1,20 +1,63 @@
+import { OverviewRouteFragmentContainer as OverviewRoute } from "./Routes/Overview"
+import { WorksRouteFragmentContainer as WorksRoute } from "./Routes/Works"
+
+graphql`
```
?

sweir27 replied:
This is needed in the `render()` function below in order to determine whether we need to redirect or not.

damassi commented on src/Apps/Artist/Routes/Works/index.tsx:
```diff
@@ -0,0 +1,145 @@
+  artist: Works_artist
+}
+
+export class WorksRoute extends Component<WorksRouteProps> {
```
`React.FC`

sweir27 commented on src/Apps/Artist/Routes/Overview/index.tsx:
```diff
@@ -36,16 +41,26 @@ export class OverviewRoute extends React.Component<OverviewRouteProps, {}> {
+  @track<OverviewRouteProps>(props => ({
+    action_type: Schema.ActionType.Click,
+    subject: "Clicked browse works for sale",
+    destination_path: props.artist.href,
```
TODO: This should probably be `/artist/${artistSlug}/works-for-sale`

sweir27 commented on src/Apps/Order/__tests__/OrderApp.test.tsx:
```diff
@@ -67,7 +67,7 @@ describe("OrderApp routing redirects", () => {
   })
 
-  it("redirects to the status route if the order is not pending", async () => {
+  it.only("redirects to the status route if the order is not pending", async () => {
```
TODO: oops

sweir27 commented:
Forgot to mention that after this PR is in, I'll follow up with another to refactor.

Right now we still have `CV`, `Shows` and `Articles` as "Routes" under the `Artist` folder.

sweir27 commented:
Agh my commit history got all mixed up. Will fix tomorrow!

sweir27 commented:
`Shop works for sale` button:
![image](https://user-images.githubusercontent.com/2081340/71193022-5f9f5c80-2257-11ea-9243-c62ac6a59e5e.png)

And on mobile
![image](https://user-images.githubusercontent.com/2081340/71193097-878ec000-2257-11ea-9035-f95bed912b93.png)

sepans commented on src/Apps/Artist/Routes/Overview/index.tsx:
```diff
@@ -94,9 +115,17 @@ export class OverviewRoute extends React.Component<OverviewRouteProps, {}> {
-              </Box>
+              <Media greaterThan="xs">
+                <Box pl={2}>
+                  <SelectedCareerAchievements artist={artist} />
```
The difference is only padding?

mzikherman commented on src/Apps/Artist/Components/ArtistHeader.tsx:
```diff
@@ -34,6 +35,16 @@ const H1 = styled.h1`
 
 const H2 = H1.withComponent("h2")
 
+export const WorksForSaleButton = styled.div`
```
Yea, should be using the Palette primitives over `div` and things like that.

mzikherman commented on src/Apps/Artist/Components/ArtistHeader.tsx:
```diff
@@ -106,25 +134,39 @@ export class LargeArtistHeader extends Component<Props> {
+                <Button
+                  variant="primaryWhite"
+                  size="small"
+                  onClick={() => this.handleBrowseWorksClick(props.artist.slug)}
```
This came up with client side routing (and we were discussing with @joeyAghion in lunch and learn), but we should try to avoid using custom click handlers, to handle navigation and routing when possible.

This feels like a link to the /works-for-sale tab of an artist page, yet uses `window.location.assign` in an `onClick` (which will break client-side routing), rather than an anchor tag with `href` pointing to that (for users w/o javascript).

sweir27 replied:
Oh!

I did this so that we could get custom tracking around the button. Is there a solution that satisfies both of these needs?

mzikherman replied:
You can still have a no-op `onClick` just for analytics tracking purposes, but the component should probably just be a [styled] anchor tag with `href`.

sweir27 commented on src/Apps/Artist/Routes/Overview/index.tsx:
```diff
@@ -94,9 +115,17 @@ export class OverviewRoute extends React.Component<OverviewRouteProps, {}> {
-              </Box>
+              <Media greaterThan="xs">
+                <Box pl={2}>
+                  <SelectedCareerAchievements artist={artist} />
```
OMG I totally forgot about that! You are so right.

mzikherman commented on src/Apps/Artist/__tests__/routes.test.tsx:
```diff
@@ -0,0 +1,181 @@
+import { routes_ArtistTopLevelQueryRawResponse } from "__generated__/routes_ArtistTopLevelQuery.graphql"
```
Very nice tests!

sweir27 commented on src/Apps/Artist/Components/NavigationTabs.tsx:
```diff
@@ -80,16 +83,50 @@ export class NavigationTabs extends React.Component<Props> {
+  related,
+  biographyBlurb,
+}: {
+  statuses: { shows: boolean; cv: boolean; articles: boolean }
```
I don't believe so (removing them makes the type `any`). I could pull in specific types from the fragment, but this also makes it very clear which arguments are needed which I think is helpful.

damassi commented:
@mzikherman / @sweir27  - can y'all elaborate on the relay cache issue mentioned in the description?

mzikherman commented:
@damassi There were two separate issues. One is the expected warnings about edges lacking an `id`, which is a Relay runtime warning.

At the same time @sweir27 was also experiencing issues where the Relay compiler was error'ing out, but the error message didn't make any sense. That was:

```
ERROR:
Parse error: TypeError: Cannot read property 'length' of undefined in […]
```

and the solution was removing the compiler cache
```
$ rm -rf $TMPDIR/RelayFindGraphQLTags-*
```

courtesy of @alloy in https://artsy.slack.com/archives/C2TQ4PT8R/p1572959259008600

alloy commented:
![](https://user-images.githubusercontent.com/2081340/70960281-99504780-204c-11ea-895f-eba4954a8cf4.png)

This warning is I believe a Relay bug. It doesn’t negatively affect things, because these edges never have IDs, but it also shouldn’t complain about it. Iirc, I wrote some of my thoughts on it in the ‘reaction mpv2 migration’ doc, so somebody should take some time to debug and trace the issue in relay-runtime and if indeed a bug submit an upstream patch.

damassi commented:
Thanks @alloy, @mzikherman  - will give that a look 👍

zephraph commented on src/Apps/Artist/ArtistApp.tsx:
```diff
@@ -26,21 +37,52 @@ export interface ArtistAppProps {
-          <ArtistHeader artist={artist} />
-        </Col>
-      </Row>
+      {route.displayNavigationTabs && (
```
Why is the navigation conditionally rendered?

sweir27 replied:
Based on the designs, we either show the navigation tabs or a back-link.

sweir27 commented:
@zephraph @mzikherman @sepans @lilyfromseattle 

Mind giving this another pass? The main difference between the last version and this one is:
- Previously, I had removed the `/cv`, `/shows`, and `/articles` routes altogether. The desire is to _keep_ them, but just remove the tabs.
- Those ☝️ sub pages only show a back link instead of tabs, so I made that change as well.
- Added a couple of modules on the overview page:
  - Works for Sale (pulls a few artworks and sticks them in a standard carousel)
  - Current shows
  - Articles

sweir27 commented:
Check out artist-big-bang.artsy.net to actually play around!

Some screenshots:
![image](https://user-images.githubusercontent.com/2081340/72826420-cf35ac80-3c46-11ea-9530-418e13a13ed8.png)


![image](https://user-images.githubusercontent.com/2081340/72826448-da88d800-3c46-11ea-9e10-fb9e5de251c2.png)

zephraph commented:
Is the `browse all works for sale` button supposed to be full width on desktop?

sepans commented on src/Apps/Order/OrderApp.tsx:
```diff
@@ -22,7 +22,7 @@ declare global {
 }
 
-const findCurrentRoute = ({ routes, routeIndices }: Match) => {
+export const findCurrentRoute = ({ routes, routeIndices }: Match) => {
```
#minor Worth moving this into a helper file?

mzikherman commented on src/Apps/Artist/Routes/Works/index.tsx:
```diff
@@ -0,0 +1,142 @@
+  const { artist } = props
+  const { sidebarAggregations } = artist
+
+  const isClient = typeof window !== "undefined"
```
This isn't great since it hardcodes `window`, isn't portable to RN, etc.

Instead, should use a `componentDidMount` if a class method, or a hook otherwise.

sweir27 replied:
👍this follows existing patterns in the artist app so will follow up afterwards and fix other instances too.

sweir27 commented on src/Apps/Artist/Components/ArtistHeader.tsx:
```diff
@@ -115,25 +145,43 @@ export class LargeArtistHeader extends Component<Props> {
+                }}
+              />
+              <WorksForSaleButton pl={4}>
+                <Link
```
Ahh! That's awesome-- didn't know about the `RouterLink`. Thanks!

zephraph commented on src/Apps/Artist/Routes/Overview/Components/WorksForSaleRail.tsx:
```diff
@@ -0,0 +1,124 @@
+}> = ({ artistID }) => {
+  const { relayEnvironment } = useContext(SystemContext)
+  return (
+    <QueryRenderer<WorksForSaleRailRendererQuery>
```
I would be interested in what we could do to help educate / catch instances of this earlier. 

Could we add a warning lint rule?

mzikherman replied:
good call, I think you may have brought up before too (when this component was first added).

I think a lint rule makes sense! I guess we're saying that we'd pretty much always prefer `SystemQueryRenderer` and so should have a lint error if someone imports `QueryRenderer`?

mzikherman commented:
High-level meta comment, which only just occurred to me as I was looking over this PR today, which I'll post for posterity, although I'm not sure how applicable this is now.

Basically, this seems like a large change to the current production artist app, and is being done in one shot, which requires a large PR, review app, QA'ing. Also, when ready, it seems like this will replace the current app, so wouldn't be AB tested?

Below is how I think I might approach this, with an eye on making some of those steps easier (and possibly enabling an AB test).

## Altered sequence proposal

- in Reaction, duplicate the `Artist` app into `Artist2`. Then, all updates could be made into `Artist2`. Since this is a new app that wouldn't replace anything, there could then be a series of smaller/WIP PR's here.

- in the Force artist app, pull in `Artist2` instead of `Artist` from Reaction based on some conditional (maybe a lab feature?).

- eventually, the conditional above could change to a proper AB test (if we want to test this)

- in Reaction, once we resolve the AB test and assuming we want to continue with the `Artist2` version, we can delete `Artist` completely and rename `Artist2` to `Artist`.

Again, apologies for the really late comments/feedback. I'm sure I'm missing a bunch of nuance in the above (when you duplicate the `Artist` app initially in Reaction, prob need to append a `2` to the Relay queries, etc.) or if you'd already considered it.

sweir27 commented:
Thanks @mzikherman!

In general I agree with you about incremental rollouts, AB tests, etc.

This is sort of a special case. 😅We'd decided not to AB test because we were confident in wanting to ship the new version as it matches our vision for what the artist page should be.

The original version just moved content from the tabs onto the overview page, so doing it in one go made a bit more sense. However, after viewing it in the review app, we realized that we needed to revisit the drawing board and re-think the design. 

I wanted to put the ✨ new vision ✨ up in a review app quickly so that we can see if we're confident in how it looks (didn't take much work given it was based on prior work + the patterns that you all have set up in this project 😄).

sweir27 commented on src/Apps/Artist/Routes/Overview/Components/WorksForSaleRail.tsx:
```diff
@@ -0,0 +1,124 @@
+const WorksForSaleRail: React.FC<WorksForSaleRailProps & {
+  onArtworkClicked: () => void
+}> = ({ artist, onArtworkClicked }) => {
+  const { user, mediator } = useContext(SystemContext)
```
It gets passed into `FillWidthItem` for the save button (perhaps to control a potential auth flow?)

mzikherman replied:
Ah, likely. Cool!

mzikherman commented:
Thanks for the explanation @sweir27 . With an AB test, I think you'd probably have to do it in the slightly more redundant/incremental approach, in order to be able to actually have both versions existing side by side.

But without the need for that, and with the ease today of composing and moving these components around, this is fine too! 👍

sweir27 commented on src/Apps/Artist/Components/ArtistCollectionsRail/ArtistCollectionEntity.tsx:
```diff
@@ -37,7 +37,14 @@ export class ArtistCollectionEntity extends React.Component<CollectionProps> {
       artworksConnection,
     } = this.props.collection
-    const artworks = artworksConnection.edges.map(({ node }) => node)
+    const artworks = get(artworksConnection, ac =>
```
Added this to fix the crash loop we're seeing due to the current elasticsearch issues on staging (would be an error on prod too, so good to fix!)

sweir27 commented on src/Apps/Artist/ArtistApp.tsx:
```diff
@@ -45,21 +59,57 @@ export const ArtistApp: React.FC<ArtistAppProps> = props => {
-          <ArtistHeader artist={artist} />
-        </Col>
-      </Row>
+      {route.displayNavigationTabs && (
```
We store some metadata (`displayNavigationTabs`) on the route so we can use it here to decide whether or not to display the "Back to artist" link or the full set of tabs.

sweir27 commented on src/Apps/Artist/ArtistApp.tsx:
```diff
@@ -87,6 +137,15 @@ export const ArtistApp: React.FC<ArtistAppProps> = props => {
   )
 }
 
+export const StyledLink = styled(RouterLink)`
```
This gets re-used throughout the artist app, which is why I chose to export it from here.

We use `RouterLink` so we can avoid full page refreshes when moving between different sub-artist-pages.

sweir27 commented on src/Apps/Artist/Components/ArtistCollectionsRail/ArtistCollectionsRail.tsx:
```diff
@@ -42,16 +43,22 @@ export class ArtistCollectionsRail extends React.Component<
-          <Sans size="3" weight="medium">
-            Browse by iconic collections
+
+          {includeTopSpacer && <Separator my={3} />}
```
Doesn't feel great to have to add this here, but we don't want to show the separator/spacer unless content _loads_, and we only know that from within this element (otherwise we'd have to fetch for data in the parent and avoid any performance benefits we'd get from loading client-side).

sweir27 commented on src/Apps/Artist/routes.tsx:
```diff
@@ -24,36 +70,51 @@ export const routes: RouteConfig[] = [
+          (artist.insights && artist.insights.length > 0)
+        const hasArtistContent = hasOverviewContent(artist)
+
+        const alreadyAtWorksForSalePath = match.location.pathname.includes(
```
This is where we determine our zero-state routing logic.

If the artist has no "overview-worthy" content, we redirect to the `/works-for-sale` path _unless_ you're already at that path (to avoid infinite looping).

sweir27 commented on src/Components/v2/SelectedCareerAchievements.tsx:
```diff
@@ -100,26 +94,13 @@ export class SelectedCareerAchievements extends React.Component<
       return null
     }
 
-    const Container = props => {
```
Found this wasn't used anywhere and added some complexity, so removed for now.

artsyit commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in v25.5.0 :rocket:
